src/dataflow/src/source/kinesis.rs

```diff
@@ -37,126 +41,237 @@ lazy_static! {
-pub fn kinesis<G>(
-    config: SourceConfig<G>,
-    connector: KinesisSourceConnector,
-) -> (
-    Stream<G, SourceOutput<Vec<u8>, Vec<u8>>>,
-    Option<SourceToken>,
-)
-where
-    G: Scope<Timestamp = Timestamp>,
-{
-    // Putting source information on the Timestamp channel lets this
-    // Dataflow worker communicate that it has created a source.
-    let ts = if let Consistency::BringYourOwn(_) = config.consistency {
-        if config.active {
-            let prev = config.timestamp_histories.borrow_mut().insert(
-                config.id.clone(),
-                TimestampDataUpdate::BringYourOwn(HashMap::new()),
+/// Contains all information necessary to ingest data from Kinesis
+pub struct KinesisSourceInfo {
+    /// Source Name
+    name: String,
+    /// Unique Source Id
+    id: SourceInstanceId,
+    /// Field is set if this operator is responsible for ingesting data
+    is_activated_reader: bool,
+    /// Kinesis client used to obtain records
+    kinesis_client: KinesisClient,
+    /// The name of the stream
+    stream_name: String,
+    /// The set of active shards
+    shard_set: HashSet<String>,
+    /// A queue representing the next shard to read from. This is necessary
+    /// to ensure that all shards are read from uniformly
+    shard_queue: VecDeque<(String, Option<String>)>,
+    /// The time at which we last refreshed metadata
+    /// TODO(natacha): this should be moved to timestamper
+    last_checked_shards: Instant,
+    /// Storage for messages that have not yet been timestamped
+    buffered_messages: VecDeque<SourceMessage<Vec<u8>>>,
+    /// Count of processed message
+    processed_message_count: i64,
+}
+
+impl SourceConstructor<Vec<u8>> for KinesisSourceInfo {
+    fn new(
+        source_name: String,
+        source_id: SourceInstanceId,
+        active: bool,
+        _worker_id: usize,
+        _worker_count: usize,
+        _consumer_activator: Arc<Mutex<SyncActivator>>,
+        connector: ExternalSourceConnector,
+        consistency_info: &mut ConsistencyInfo,
+        _encoding: DataEncoding,
+    ) -> Result<Self, failure::Error> {
+        let kc = match connector {
+            ExternalSourceConnector::Kinesis(kc) => kc,
+            _ => unreachable!(),
+        };
+
+        let state = block_on(create_state(kc, &source_name, source_id, consistency_info));
+        match state {
+            Ok((kinesis_client, stream_name, shard_set, shard_queue)) => Ok(KinesisSourceInfo {
+                name: source_name,
+                id: source_id,
+                is_activated_reader: active,
+                kinesis_client,
+                shard_queue,
+                last_checked_shards: Instant::now(),
+                buffered_messages: VecDeque::new(),
+                shard_set,
+                stream_name,
+                processed_message_count: 0,
+            }),
+            Err(e) => Err(failure::err_msg(e.to_string())),
+        }
+    }
+}
+
+impl KinesisSourceInfo {
```

Comment:
Are the other sources having their functions implemented on their `SourceInfo` struct?

Reply:
Yes